client/core/core.go

```diff
@@ -591,6 +591,137 @@ func (dc *dexConnection) compareServerMatches(srvMatches map[order.OrderID]*serv
+//   observes that there are no active matches for the trades.
+// - coins are unlocked either as the affected trades' matches are swapped or
+//   revoked (for trades with active matches), or when the trades are retired.
+func (dc *dexConnection) reconcileTrades(serverOrders []*msgjson.Order) (unknownOrdersCount, reconciledOrdersCount int) {
+	dc.tradeMtx.RLock()
+	// Check for unknown orders reported as active by the server. If such
+	// exists, could be that they were known to the client but were thought
+	// to be inactive and thus were not loaded from db or were retired.
+	serverMsgOrders := make(map[order.OrderID]*msgjson.Order, len(serverOrders))
+	for _, msgOrder := range serverOrders {
+		var oid order.OrderID
+		copy(oid[:], msgOrder.OrderID)
+		if _, tracked := dc.trades[oid]; tracked {
+			serverMsgOrders[oid] = msgOrder
+		} else {
+			log.Errorf("Unknown order %v reported by DEX %s as active", oid, dc.acct.host)
+			unknownOrdersCount++
+		}
+	}
+	knownActiveTrades := make(map[order.OrderID]*trackedTrade)
+	for oid, trade := range dc.trades {
+		trade.mtx.RLock()
+		if trade.metaData.Status == order.OrderStatusEpoch || trade.metaData.Status == order.OrderStatusBooked {
+			knownActiveTrades[oid] = trade
+		} else if serverMsgOrder := serverMsgOrders[oid]; serverMsgOrder != nil {
+			log.Warnf("Inactive order %v, status %s reported by DEX %s as active, status %s",
+				oid, trade.metaData.Status, dc.acct.host, order.OrderStatus(serverMsgOrder.Status))
+		}
+		trade.mtx.RUnlock()
+	}
+	dc.tradeMtx.RUnlock()
+
+	updateOrder := func(trade *trackedTrade, msgOrder *msgjson.Order) {
+		reconciledOrdersCount++
+		oid := trade.ID()
+		previousStatus := trade.metaData.Status
+		newStatus := order.OrderStatus(msgOrder.Status)
+		trade.Trade().SetFill(msgOrder.Fill)
+		trade.metaData.Status = newStatus
+		if err := trade.db.UpdateOrder(trade.metaOrder()); err != nil {
+			log.Errorf("Error updating status in db for order %v from %v to %v", oid, previousStatus, newStatus)
+		} else {
+			log.Debugf("Order %v updated from recorded status %v to new status %v reported by DEX %s",
+				oid, previousStatus, newStatus, dc.acct.host)
+		}
+	}
+
+	// Compare the status reported by the server for each known active trade. Orders
+	// for which the server did not return a status are no longer active (now Executed,
+	// Canceled or Revoked). Use the order_status route to determine the correct status
+	// for such orders and update accordingly.
+	var orderStatusRequests []*msgjson.OrderStatusRequest
+	for oid, trade := range knownActiveTrades {
+		serverMsgOrder := serverMsgOrders[oid]
+		if serverMsgOrder == nil {
+			// Order not returned by server. Must be inactive now.
+			// Request current status from the DEX.
+			orderStatusRequests = append(orderStatusRequests, &msgjson.OrderStatusRequest{
+				Base:    trade.Base(),
+				Quote:   trade.Quote(),
+				OrderID: oid[:],
+			})
```

Comment:
I've not quite figured it out yet, but there's a problem here. If appending more than 1 request, the value of the last appended request overrides the values of all previously appended requests. I'm constantly seeing 3 items in the slice but all with same value as the last appended item. So instead of getting statuses for 3 orders, I only get for one. Will push a fix once I figure it out.

Reply:
Turns out it was because I was using the loop variable oid. Replaced with trade.ID().Bytes() and now all good. Still wondering why it's safe to use the second loop variable (trade) but not oid.